python_src/models/police_report.py

```diff
@@ -0,0 +1,251 @@
+from __future__ import annotations
+from typing import Tuple
+from abc import ABC, abstractmethod
+
+CAT_A = 'A'
+CAT_B = 'B'
+CAT_C = 'C'
+CAT_D = 'D'
+
+CAT_A_DATA = ["ANY"]
+CAT_B_DATA = ['F', 'M', 'B']
+CAT_C_DATA = ['L', 'R']
+CAT_D_DATA = ["FL", "FR", "ML", "MR", "BL", "BR"]
+
+CATEGORIES = [{"type": CAT_A, "data": CAT_A_DATA},
+              {"type": CAT_B, "data": CAT_B_DATA},
+              {"type": CAT_C, "data": CAT_C_DATA},
+              {"type": CAT_D, "data": CAT_D_DATA}]
+
+NO_CRASH = 0
+
+
+class Creator(ABC):
+    @abstractmethod
+    def factory_method(self):
+        pass
+
+    def match_operation(self, outputs: list, targets: list) -> Tuple[int, int, int]:
+        police_report = self.factory_method()
+        return police_report.process(outputs, targets)
+
+
+"""
+Concrete Creators override the factory method in order to change the resulting
+report's type.
+"""
+
+
+class ConcreteCreatorA(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeA()
+
+
+class ConcreteCreatorB(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeB()
+
+
+class ConcreteCreatorC(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeC()
+
+
+class ConcreteCreatorD(Creator):
+    def factory_method(self) -> Report:
+        return ReportTypeD()
```

Comment:
The names of those classes are not very descriptive. Can we rename them? 
AnyCreator, CarSideCreator or SideCreator, etc..?
Maybe we can add a line of documentation on what they do?

BTW, what does it mean Creator? PoliceReport creator?

Reply:
I rename them to AnyCreator, SideCreator, ComponentCreator and ComponentSideCreator, some comments were added to tell people what they do. [Sub]Creators override the base factory method from the ReportCreator interface so it returns a different type of report.